### PR TITLE
Format hash directives inside other hash directives of complex trivia

### LINKED ISSUE #285: #else not formatted in #if'd out code
Ported from TFS WorkItem: <b>1089196</b>

------------------------------

<p><b>Repro Steps:</b></p>
<P><B>Affected Branch:</B> <BR><B>Affected Build:</B> <BR><B>Affected Product Language:</B> <BR><BR><B>Steps to reproduce:</B><BR><BR>1. Copy/paste C#:</P>
<P>&nbsp;</P><PRE style="FONT-SIZE: 16px; FONT-FAMILY: Consolas; BACKGROUND: white; COLOR: hotpink"><SPAN style="COLOR: blue">class</SPAN> <SPAN style="COLOR: #2b91af">MyClass</SPAN>
<SPAN style="COLOR: black">{</SPAN>
    <SPAN style="COLOR: blue">static</SPAN> <SPAN style="COLOR: blue">void</SPAN> <SPAN style="COLOR: black">Main</SPAN><SPAN style="COLOR: black">(</SPAN><SPAN style="COLOR: blue">string</SPAN><SPAN style="COLOR: black">[]</SPAN> <SPAN style="COLOR: black">args</SPAN><SPAN style="COLOR: black">)</SPAN>
    <SPAN style="COLOR: black">{</SPAN>
<SPAN style="COLOR: blue">#if</SPAN> <SPAN style="COLOR: blue">false</SPAN>
<SPAN style="COLOR: gray">$$
 
</SPAN><SPAN style="COLOR: blue">#endif</SPAN>
    <SPAN style="COLOR: black">}</SPAN>
<SPAN style="COLOR: black">}</SPAN></PRE>
<P>&nbsp;</P>
<P>2. Remove the $$ if necessary and type <STRONG>&lt;enter&gt;#else</STRONG></P>
<P>&nbsp;</P>
<P><STRONG>Expected</STRONG>: #else moved to left alignment</P>
<P><STRONG>Actual</STRONG>: #else is still indented at method body level</P>

------------------------------

<p><b>Revisions:</b></p>

1) Created By David Poeschl (12/1/2014 1:43:39 PM)

------------------------------

2) Edited By David Poeschl (12/1/2014 1:44:00 PM)

<p>Note: this may be related to #286 due to parse tree shape</p>

------------------------------

3) Edited By Kevin Pilch-Bisson (12/2/2014 8:54:54 AM)

<p>1. I don't think this is related to #286, since in C# you already have a &quot;Bad directive trivia&quot; once you have typed the &quot;#&quot;.</p>
<p>&nbsp;</p>
<p>2. This works if you type it on the line immediately after the #if false, but not on subsequent lines.</p>
<p>&nbsp;</p>
<p>3. Format document doesn't fix it either.</p>

------------------------------

### PR DESCRIPTION
**Pull request by basoundr**
Fix #285 : Analyze DisabledTextTrivia appropriately, so that when it is followed by a hash directive, we will be able to determine if we need to format the hash directive.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs [hunk 1]
```diff
@@ -300,6 +297,17 @@ private bool ShouldFormat()
...
 
+            private bool OnDisabledTextTrivia(SyntaxTrivia trivia, int index)
+            {
+                if (trivia.IsKind(SyntaxKind.DisabledTextTrivia) &&
+                    SyntaxFacts.IsNewLine(trivia.ToString().Last()))
+                {
+                    ResetStateAfterNewLine(index);
+                }
+
+                return false;
```
Comment (balajikris) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: this function always returns false. So the conditional for OnDisabledTextTrivia() in ShouldFormat() is never true.

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs [hunk 2]
```diff
@@ -5913,5 +5913,100 @@ void M()
...
+
+#else
+#
+
+#endif
+    }
+}";
+            AssertFormat(expected, code);
+        }
     }
```
Comment (balajikris) [inline src/Workspaces/CSharpTest/Formatting/FormattingTests.cs]: optional: tests with **nested** `#if`statements and `#elif`, if they are relevant.

File: src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs (hunk 1 above)
Comment (basoundr) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: `OnDisabledTextTrivia()` like the other methods in `ShouldFormat()` is a method which introduces side effects. This is the intended behavior of the method.

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs (hunk 2 above)
Comment (basoundr) [inline src/Workspaces/CSharpTest/Formatting/FormattingTests.cs]: These conditions are irrelevant to the changes I made.

File: src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs (hunk 1 above)
Comment (balajikris) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: Oh I see. So, it adjusts position of something and says don't format this again.

**Comment by balajikris** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs (hunk 1 above)
Comment (basoundr) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: `CodeShapeAnalyzer` is a StateMachine and `OnDisabledTextTrivia()` just modifies the state.

**Comment by basoundr** [comment]
@heejae, a sign off from you and I will merge the PR

File: src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs [hunk 3]
```diff
@@ -300,6 +297,17 @@ private bool ShouldFormat()
                 return false;
             }
 
+            private bool OnDisabledTextTrivia(SyntaxTrivia trivia, int index)
+            {
+                if (trivia.IsKind(SyntaxKind.DisabledTextTrivia) &&
+                    SyntaxFacts.IsNewLine(trivia.ToString().Last()))
```
Comment (heejaechang) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: is last access last char directly? or enumerate through all chars to the end?

Comment (heejaechang) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: it is really unfortunate that only way to check last char of a trivia is ToString.

**Comment by heejaechang** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs (hunk 3 above)
Comment (basoundr) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: Last() is an extension method which gets the last char. I should introduce a NullEmpty check.

Comment (heejaechang) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: I don’t what it does, what I mean is, is it internally do string[string.length -1] or is it do

Foreach(var c in string)
{
… get to the last one.
Return ch;
}


-          Heejae

Comment (basoundr) [inline src/Workspaces/CSharp/Portable/Formatting/Engine/Trivia/TriviaDataFactory.CodeShapeAnalyzer.cs]: It does string[string.length -1]. Hence I added IsNullOrEmpty check.

**Pull request merged**